### PR TITLE
rework how allow-missing-images resolves

### PR DESCRIPTION
**Pull request by bparees**
fixes https://github.com/openshift/origin/issues/4991

### CODE REVIEW

**Comment by bparees** [comment]
[test]

**Comment by bparees** [comment]
@csrwng ptal

**Comment by csrwng** [comment]
test/cmd/newapp.sh is failing

**Comment by csrwng** [comment]
@bparees - this looks good, pending a fix to the test. While re-familiarizing myself with this code, a saw a couple of things:
- This is a bug:
https://github.com/bparees/origin/blob/allow_missing/pkg/generate/app/componentresolvers.go#L118-L119
It should check whether the err is an instance of ErrMultipleMatches and add the multiple matches to the list of candidates.

**Comment by csrwng** [comment]
And the resolve getting called with different sets of components

**Comment by bparees** [comment]
> This is a bug: https://github.com/bparees/origin/blob/allow_missing/pkg/generate/app/componentresolvers.go#L118-L119 It should check whether the err is an instance of ErrMultipleMatches and add the multiple matches to the list of candidates.

@csrwng fixed and test added, ptal.

File: pkg/generate/app/componentresolvers_test.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+		results = append(results, &ComponentMatch{Argument: fmt.Sprintf("match%d", i), Score: 0.0})
+	}
+
+	return results, nil
+}
+
+func TestWeightedResolvers(t *testing.T) {
+	resolver1 := WeightedResolver{mockSearcher{2}, 1.0}
+	resolver2 := WeightedResolver{mockSearcher{3}, 1.0}
+	wr := WeightedResolvers{}
```
Comment (csrwng) [inline pkg/generate/app/componentresolvers_test.go]: Just an observation... it'd be more succinct to say
```
wr := WeightedResolvers{ resolver1, resolver2 }
```

Comment (bparees) [inline pkg/generate/app/componentresolvers_test.go]: updated.

File: test/cmd/newapp.sh [hunk 2]
```diff
@@ -81,9 +81,25 @@ oc delete template ruby-helloworld-sample
...
+	rm -rf $tmp
+	[ "$(oc new-app  test/scratchimage~https://github.com/openshift/ruby-hello-world.git --strategy=docker -o yaml |& tr '\n' ' ' | grep -E "from:\s+kind:\s+DockerImage\s+name:\s+test/scratchimage:latest\s+")" ]
+fi
+
+# do not allow use of non-existent image (should fail)
 [ "$(oc new-app  openshift/bogusImage https://github.com/openshift/ruby-hello-world.git -o yaml 2>&1 | grep "no image or template matched")" ]
-# TODO: Fix how --allow-missing-images is handled when no docker daemon present
-# [ "$(oc new-app  openshift/bogusImage https://github.com/openshift/ruby-hello-world.git -o yaml --allow-missing-images)" ]
+# allow use of non-existent image (should succeed)
+[ "$(oc new-app  openshift/bogusImage https://github.com/openshift/ruby-hello-world.git -o yaml --allow-missing-images)" ]
```
Comment (csrwng) [inline test/cmd/newapp.sh]: Isn't this the same test as in line 84?

**Comment by csrwng** [comment]
Apart from test comment above, LGTM

File: test/cmd/newapp.sh (hunk 2 above)
Comment (bparees) [inline test/cmd/newapp.sh]: yes, cut+paste issue.  fixed.

**Comment by bparees** [comment]
[merge]

File: test/cmd/newapp.sh [hunk 3]
```diff
@@ -81,9 +81,23 @@ oc delete template ruby-helloworld-sample
 oc new-app https://github.com/openshift/ruby-hello-world -l app=ruby
 oc delete all -l app=ruby
 
-# allow use of non-existent image
+if `hash docker > /dev/null 2>&1`; then
```
Comment (smarterclayton) [inline test/cmd/newapp.sh]: This really shouldn't be here and should be in an integration test.  test/cmd should *not* prereq Docker.  Please open a follow up.

Comment (liggitt) [inline test/cmd/newapp.sh]: Won't that fail on Travis?

Comment (bparees) [inline test/cmd/newapp.sh]: @csrwng promised me that travis didn't have docker at all, which is why i added the check as i did.

Comment (bparees) [inline test/cmd/newapp.sh]: and @smarterclayton it doesn't pre-req docker, it uses docker if it's available, otherwise it skips the test.

**Pull request merged**